minerl/env/core.py

```diff
@@ -175,31 +171,6 @@ def _robust_launch_new_instance(self, *, max_tries=3) -> InstanceManager:
         raise RuntimeError(f"Failed to build and launch Minecraft instance "
                            f"{max_tries} times. Giving up.")
 
-    def _setup_spaces(self, observation_space, action_space):
-        self.action_space = action_space
-        self.observation_space = observation_space
-
-        def map_space(space):
```

Comment:
I'm a little confused about what this `map_space` function was doing; was it just duplicating the no_op action for each space? If so, why does it manually specify a value rather than just calling the no_op function directly? I just want to confirm this wasn't doing something subtle we're not realizing as valuable before we delete it.

Reply:
Double-checking this seems like a good idea.

I'm 95% sure that `map_space` just reimplements what is currently called [the `.no_op` method ](https://github.com/HumanCompatibleAI/minerl/blob/master/minerl/env/spaces.py#L42)in every custom Space, and that the result of `map_space` is saved as `minerl_env._default_action` to be returned via the dynamically set `minerl_env.action_space.noop()` action (which has to close over `minerl_env` in order to return `minerl_env._default_action`).

Reply:
I believe the engineering reason behind having `map_space` save to `minerl_env._default_action` and have `noop()` return a deepcopy of `minerl_env._default_action` is so that at initialization that MineRLEnv can have a custom noop action via the `noop_action` parameter.

Reply:
(Nothing uses this custom noop action right now AFAICT with Pycharm searching for `noop_action`)

Reply:
Oh, this is prob relevant for understanding -- we only ever call `map_space` if the custom noop action wasn't set https://github.com/HumanCompatibleAI/minerl/blob/master/minerl/env/core.py#L193

Reply:
Yeah, I think my question is something like: if this is the same as no_op in every space, why don't they just assign the result from the space's no_op() function  to `_default_action` if there is no default action passed in?

Reply:
I don't know.

Maybe the author forgot that there was the other implementation before they coded this in? Or maybe `map_spaces` is a copy of some original implementation from Malmo. There's an [unreached line](https://github.com/HumanCompatibleAI/minerl/blob/master/minerl/env/core.py#L189) in `map_spaces` referencing a method that doesn't seem to exist in any MineRL space or standard Gym space, which seems to support the "copied from somewhere else" hypothesis.

Reply:
🤦‍♀ Yeah, a frustrating thing about MineRL being occasionally such a kludge of a codebase (see also: an entire top level directory of source that was just never called, but left confusingly lying around) is that it's often hard to know how hard to chesterton's fence on it. 

Could we do a quick empirical check by temporarily bringing back the `map_space` code base and ensuring the no_ops are in fact empirical for some set of action spaces and observation spaces in existing MineRL environments?

Reply:
Sure, I'll use the current `master` to compare outputs from `no_op` and `noop`.

Reply:
Confirmed that `no_op` and `noop` are the same for four arbitrary envs (containing Box, Enum, and Discrete action spaces).

https://gist.github.com/shwang/87176375e62e71dad0488f1ce1b00fca

Reply:
Awesome, thanks for doing that.